src/qmapshack/poi/CPoiPOI.cpp

```diff
@@ -86,126 +88,269 @@ void CPoiPOI::draw(IDrawContext::buffer_t& buf)
-                   !loadedPOIs[category][minLonM10].contains(minLatM10))
+                if(!loadedPOIsByArea.contains(categoryID) ||
+                   !loadedPOIsByArea[categoryID].contains(minLonM10) ||
+                   !loadedPOIsByArea[categoryID][minLonM10].contains(minLatM10))
                 {
-                    loadPOIsFromFile(category, minLonM10, minLatM10);
+                    loadPOIsFromFile(categoryID, minLonM10, minLatM10);
                 }
-                for(const rawPoi_t& poiItemFound : loadedPOIs[category][minLonM10][minLatM10])
+                for(quint64 poiFoundID : loadedPOIsByArea[categoryID][minLonM10][minLatM10])
                 {
-                    //Maybe look through the whole code of selecting items from a map to avoid this conversion
-                    if(degRect.contains(poiItemFound.coordinates * RAD_TO_DEG))
+                    const CRawPoi& poiItemFound = loadedPois[poiFoundID];
+                    if(!copiedItems.contains(poiItemFound.getKey()))
                     {
-                        pois.append(poiItemFound.toPoi(category));
+                        //Maybe look through the whole code of selecting items from a map to avoid this conversion
+                        if(degRect.contains(poiItemFound.getCoordinates() * RAD_TO_DEG))
+                        {
+                            pois.insert(poiItemFound.toPoi());
+                            copiedItems.insert(poiItemFound.getKey());
+                        }
                     }
                 }
             }
         }
     }
+
+    //Find Highlights
+    for(const poiGroup_t& poiGroup:displayedPois)
+    {
+        if(degRect.contains(poiGroup.iconCenter * RAD_TO_DEG))
+        {
+            posPoiHighlight.append(poiGroup.iconCenter);
+        }
+    }
 }
 
 bool CPoiPOI::getToolTip(const QPoint &px, QString &str) const
 {
-    poi_t poiFound;
-    bool success = findPoiCloseBy(px, poiFound);
+    poiGroup_t poiGroup;
+    bool success = getPoiGroupCloseBy(px, poiGroup);
     if(success)
     {
-        str += "<b>" + poiFound.name + "</b><br>";
-        str += poiFound.desc;
+        if(poiGroup.pois.count() == 1)
+        {
+            const CRawPoi& poiFound = loadedPois[*poiGroup.pois.begin()];
+            const QString& name = poiFound.getName(false);
+            if(!name.isEmpty())
+            {
+                str += "<b>" + name + "</b><br>\n";
+            }
+            str += tr("From category <b>%1</b><br>\n").arg(poiFound.getCategory());
+            str += poiFound.getData().join("<br>\n");
+        }
+        else
+        {
+            str += tr("<i>Zoom in to see more details.</i>");
+            if(poiGroup.pois.count() <= 10)
+            {
+                str += tr("<br>\nPOIs at this point:");
+                for(quint64 poiID : poiGroup.pois)
+                {
+                    str += "<br>\n<b>" + loadedPois[poiID].getName() + "</b>";
+                }
+            }
+        }
```

Comment:
From translators point of view it would be nicer to have strings without html tags. From developers point of view it's a lot of nasty string concatenation. But I would favor the xlators POV (as I have to do it later... ;)).